### PR TITLE
Set-output-expression: Enable setting of a prior to an output node

### LINKED ISSUE #317: Randomize genome only partially
In certain use cases it would be useful to only randomize part of the genome, this is currently not possible.

### PR DESCRIPTION
**Pull request by HenrikMettler**
In some use cases it is nice to set (one of) the output(s) to a prior. One way of doing this is by inserting a dna sequence corresponding to the prior at the first internal node(s). `slice_dna` gives this possibility, without requiring the user to set the full dna. 

Marked as draft, since test coverage is minimal and code not optimized. 

(Note: I also looked at implementing the prior from a function/sympy string, however this seems quite tedious, especially if the prior contains more than 1 operation. This implementation at least circumvents the need to set the full dna :) )

closes #317 (by circumventing the issue)

### CODE REVIEW

File: cgp/genome.py [hunk 1]
```diff
@@ -270,6 +269,57 @@ def randomize(self, rng: np.random.RandomState) -> None:
         # accept generated dna if it is valid
         self.dna = dna
 
+    def slice_dna(self, dna_insert: List[int], output_node_idx: int = 0):
```
Comment (jakobj) [inline cgp/genome.py:272]: maybe `slice_dna` -> `splice`?

Comment (jakobj) [reply cgp/genome.py:272]: maybe also also allow providing a starting point which by default is 0?

File: cgp/genome.py [hunk 2]
```diff
@@ -394,10 +444,12 @@ def _replace_invalid_address_alleles(self, dna: List[int], rng: np.random.Random
                 dna[gene_idx] = gene_value
 
     def _get_addable_nodes(
-        self, node_dependencies: Dict[int, Set[int]], used_node_indices: List[int] = [],
+        self,
```
Comment (jakobj) [inline cgp/genome.py:447]: are these changes necessary for this PR?

File: cgp/genome.py [hunk 3]
```diff
@@ -413,7 +465,7 @@ def _get_region(self, region_idx: int, dna: List[int]) -> List[int]:
         ]
 
     def _determine_node_dependencies(self) -> Dict[int, Set[int]]:
-        """ Determines the set of node indices on which each node depends.
+        """Determines the set of node indices on which each node depends.
```
Comment (jakobj) [inline cgp/genome.py:468]: are these changes necessary for this PR?

File: test/test_genome.py [hunk 4]
```diff
@@ -13,7 +14,11 @@ def test_check_dna_consistency():
 
     primitives = (cgp.Add,)
     genome = cgp.Genome(
-        params["n_inputs"], params["n_outputs"], params["n_columns"], params["n_rows"], primitives,
+        params["n_inputs"],
```
Comment (jakobj) [inline test/test_genome.py:17]: are these changes necessary for this PR?

File: test/test_genome.py [hunk 5]
```diff
@@ -210,7 +215,11 @@ def test_region_iterators():
...
```
Comment (jakobj) [inline test/test_genome.py:218]: are these changes necessary for this PR?

File: test/test_genome.py [hunk 6]
```diff
@@ -766,3 +775,16 @@ def test_ncolumns_zero(rng):
 
     CartesianGraph(genome).to_func()
     CartesianGraph(genome).to_numpy()
+
+
+def test_slice_dna(genome_params, rng):
+
+    genome = cgp.Genome(**genome_params)
```
Comment (jakobj) [inline test/test_genome.py:782]: just to make sure to not get false positives you could do this 10 times

**Review by jakobj — CHANGES_REQUESTED** [review]
awesome, thanks a lot for working on this! :rocket: 

i've left a few comments, in particular i wasn't sure whether all changes are relevant for this PR. please remove those which are not.

File: cgp/genome.py (hunk 2 above)
Comment (HenrikMettler) [reply cgp/genome.py:447]: No. I think there is an error in my pre-commit hooks, will change this and revert these changes

**Comment by jakobj** [comment]
ping @HenrikMettler

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 repos:
 -   repo: https://github.com/ambv/black
-    rev: stable
+    rev: 19.10b0
```
Comment (HenrikMettler) [inline .pre-commit-config.yaml:3]: Has to be the same as in extra_requirements, otherwise there seems to be a formatting conflict.

Comment (HenrikMettler) [reply .pre-commit-config.yaml:3]: (move to seperate PR)

File: cgp/genome.py [hunk 8]
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        Returns
+        ----------
+        None
+        """
+
+        # check that dna is initialized
+        if len(self.dna) == 0:
+            raise AssertionError(
+                "Can only splice genome with initialized dna, " "use genome.randomize first"
+            )
```
Comment (HenrikMettler) [inline cgp/genome.py:299]: Can be removed, since checked in self.dna

File: cgp/genome.py [hunk 9]
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        n_insert_nodes = int(n_inserts / (self._length_per_region))
+
+        # check that insert placing is valid wrt. the number of internal nodes
+        if internal_start_node + n_insert_nodes >= self._n_hidden:
+            raise ValueError(
+                "Insert is too long to be placed at this location. "
+                "Choose smaller internal_start_node or idx"
+            )
+
+        dna = self._dna.copy()
```
Comment (HenrikMettler) [inline cgp/genome.py:318]:
```suggestion
        dna = self.dna
```

File: cgp/genome.py [hunk 10]
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        """
+
+        # check that dna is initialized
+        if len(self.dna) == 0:
+            raise AssertionError(
+                "Can only splice genome with initialized dna, " "use genome.randomize first"
+            )
+        # check that dna insert is a multiple of region length
+        if len(dna_insert) % (self._length_per_region) != 0:
+            raise AssertionError("Dna insert is not of proper length")
```
Comment (HenrikMettler) [inline cgp/genome.py:302]:
```suggestion
            raise ValueError("Dna insert is not of proper length")
```

File: cgp/genome.py [hunk 11]
```diff
@@ -270,6 +269,68 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        # check that output_node_idx is valid
+        if output_node_idx > self._n_outputs - 1:
+            raise ValueError("Output node idx too high, choose wrt to n_outputs")
+
```
Comment (HenrikMettler) [inline cgp/genome.py:307]: add test that `output_node_idx` is larger than zero.

**Comment by HenrikMettler** [comment]
I changed the logic a bit: The two parts of the previous `splice_dna` function are now split up into two separate functions (for splicing and changing the output address ). What was previously splice_dna is now called `set_prior` to better reflect what the function actually does. This function calls the other two functions once each.

File: cgp/cartesian_graph.py [hunk 12]
```diff
@@ -63,13 +63,11 @@ def __repr__(self) -> str:
         return "CartesianGraph(" + str(self._nodes) + ")"
 
     def print_active_nodes(self) -> str:
-        """Print a representation of all active nodes in the graph.
-        """
+        """Print a representation of all active nodes in the graph."""
```
Comment (jakobj) [inline cgp/cartesian_graph.py:66]: please revert all formatting changes in this PR

File: cgp/genome.py [hunk 13]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
         # accept generated dna if it is valid
         self.dna = dna
 
+    def splice_dna(self, dna_insert: List[int], internal_start_node: int = 0) -> int:
```
Comment (jakobj) [inline cgp/genome.py:272]:
```suggestion
    def splice_dna(self, new_dna: List[int], internal_start_node: int = 0) -> int:
```

File: cgp/genome.py [hunk 14]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        """Splice the dna.
+
+        Replaces a segment of the dna starting from internal start node with the dna insert
```
Comment (jakobj) [inline cgp/genome.py:275]:
```suggestion
        Replaces a segment of the dna starting from internal start node with new dna.
```

File: cgp/genome.py (hunk 13 above)
Comment (jakobj) [reply cgp/genome.py:272]: also in the following

File: cgp/genome.py [hunk 15]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        Parameters
+        ----------
+        dna_insert: List[int]
+            dna segment to be inserted.
```
Comment (jakobj) [inline cgp/genome.py:280]:
```suggestion
            New dna to be inserted.
```

File: cgp/genome.py [hunk 16]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        internal_start_node: int
```
Comment (jakobj) [inline cgp/genome.py:281]:
```suggestion
        internal_start_node: int, optional
```

File: cgp/genome.py [hunk 17]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+            idx of the internal node, where the insert starts.
```
Comment (jakobj) [inline cgp/genome.py:282]:
```suggestion
            Index of the internal node, where the insert starts.
```

File: cgp/genome.py [hunk 18]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+            Defaults to 0.
+
+        Returns
+        ----------
+        address_last_inserted_node: int
+            address of the last node inserted.
```
Comment (jakobj) [inline cgp/genome.py:288]:
```suggestion
        int
            Address of the last node for which the dna was changed.
```

File: cgp/genome.py [hunk 19]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        """
+        if len(dna_insert) % self._length_per_region != 0 or len(dna_insert) == 0:
+            raise ValueError("Length of inserted dna must be a multiple of region length")
```
Comment (jakobj) [inline cgp/genome.py:291]:
```suggestion
            raise ValueError(f"Length of inserted dna must be a multiple of region length ({self._length_per_region})")
```

provides a bit more info to the user

File: cgp/genome.py [hunk 20]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        n_inserted_nodes = len(dna_insert) // self._length_per_region
+
+        if internal_start_node + n_inserted_nodes >= self._n_hidden:
+            raise ValueError("Inserted dna too long")
```
Comment (jakobj) [inline cgp/genome.py:296]:
```suggestion
            raise ValueError("New dna too long")
```

File: cgp/genome.py [hunk 21]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        dna = self.dna
+
+        start_idx = (self._n_inputs + internal_start_node) * self._length_per_region
+        end_idx = start_idx + len(dna_insert)
+        dna[start_idx:end_idx] = dna_insert
+
+        self.dna = dna
+
+        address_last_inserted_node = self._n_inputs + internal_start_node + n_inserted_nodes - 1
```
Comment (jakobj) [inline cgp/genome.py:306]: the "address" is just the same as the index, right? so we could drop the "address_" prefix here, it doesn't provide more information. we also say "internal_start_node" and not "internal_start_node_address"

File: cgp/genome.py [hunk 22]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        return address_last_inserted_node
+
+    def change_output_address(self, address_node: int, output_node_idx: int = 0):
```
Comment (jakobj) [inline cgp/genome.py:310]:
```suggestion
    def change_address_gene_of_output_node(self, new_address: int, output_node_idx: int = 0):
```

File: cgp/genome.py [hunk 23]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        """Change the address of an output gene
```
Comment (jakobj) [inline cgp/genome.py:311]:
```suggestion
        """Change the address gene of an output node.
```

File: cgp/genome.py [hunk 24]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        Parameters
+        ----------
+        address_node: int
+            Address of the node from which the output should read
+        output_node_idx: int
+            If defined, idx of the output node for which the address node is changed.
```
Comment (jakobj) [inline cgp/genome.py:318]:
```suggestion
            Index of the output node for which the address node is changed.
```
are these indices relative (0 is the first output) or absolute (0 is the first input)? mention explicitly here please

File: cgp/genome.py [hunk 25]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+            Defaults to 0.
+
+        Returns
+        ----------
+        None
+        """
+        if address_node < 0 or address_node > self._n_hidden:
+            raise ValueError("Address node non valid")
```
Comment (jakobj) [inline cgp/genome.py:326]:
```suggestion
            raise ValueError("New address not valid; must be the index of a hidden node")
```

File: cgp/genome.py [hunk 26]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        if output_node_idx < 0:
+            raise ValueError("Output node idx must be non-negative")
```
Comment (jakobj) [inline cgp/genome.py:328]:
```suggestion
            raise ValueError("Output node index must be non-negative")
```

File: cgp/genome.py [hunk 27]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        if output_node_idx > self._n_outputs - 1:
+            raise ValueError("Output node idx must be smaller than number of outputs")
```
Comment (jakobj) [inline cgp/genome.py:330]:
```suggestion
            raise ValueError("Output node index must be smaller than number of outputs")
```

File: cgp/genome.py [hunk 28]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        dna = self.dna
+
+        idx_output_address_gene = (
+            self._n_inputs + self._n_hidden + output_node_idx
+        ) * self._length_per_region + 1  # address gene has offset 1
+        dna[idx_output_address_gene] = address_node
+        self.dna = dna
+
+    def set_prior(
```
Comment (jakobj) [inline cgp/genome.py:340]: if not sure whether "prior" is the best name. can't we just say "set_expression_for_output" or something like that?

File: cgp/genome.py [hunk 29]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        self, dna_insert: List[int], internal_start_node: int = 0, output_node_idx: int = 0
+    ):
+        """Set a prior for one output node
+
+        Replaces internal (hidden) nodes with user defined insert, starting from desired internal
+        node. Sets a definable output node to read from the last node of the insert.
```
Comment (jakobj) [inline cgp/genome.py:346]:
```suggestion
            Replaces part of the dna with user defined values starting from the specified internal
            node. Sets the specified output node to read from the last node for which dna was changed.
```

File: cgp/genome.py [hunk 30]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+
+        Parameters
+        ----------
+        dna_insert: List[int]
+            dna segment to be inserted at the first internal (hidden) nodes.
+        internal_start_node: int
+            idx of the internal node, where the insert starts.
```
Comment (jakobj) [inline cgp/genome.py:353]:
```suggestion
            index of the internal node, where the insert starts.
```

File: cgp/genome.py [hunk 31]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+            Defaults to 0.
+        output_node_idx: int
+            If defined, idx of the output node which will read the last node of the insert.
```
Comment (jakobj) [inline cgp/genome.py:356]:
```suggestion
            Index of the output node which will read the last node of the insert.
```

File: cgp/genome.py [hunk 32]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+            Defaults to 0.
+        Returns
+        ----------
+        None
+        """
+
+        # insert new dna at desired location
```
Comment (jakobj) [inline cgp/genome.py:363]:
```suggestion
```
your code and function names are so clear, no need for these comments

File: cgp/genome.py [hunk 33]
```diff
@@ -270,6 +269,105 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        address_last_inserted_node = self.splice_dna(dna_insert, internal_start_node)
+
+        # make selected output node read from last inserted node
```
Comment (jakobj) [inline cgp/genome.py:366]:
```suggestion
```

File: test/test_genome.py [hunk 34]
```diff
@@ -766,3 +767,77 @@ def test_ncolumns_zero(rng):
...
     CartesianGraph(genome).to_func()
     CartesianGraph(genome).to_numpy()
+
+
+def test_splice_dna(genome_params, rng):
+
+    genome = cgp.Genome(**genome_params)
+
+    with pytest.raises(ValueError):
+        [] = genome.splice_dna(dna_insert=[1, 2, 3])
```
Comment (jakobj) [inline test/test_genome.py:777]:
```suggestion
        genome.splice_dna(dna_insert=[1, 2, 3])
```
not sure what this does, but if you don't want to use the return value of a function you just don't assign it a name

**Review by jakobj — CHANGES_REQUESTED** [review]
yes, this split is nice, makes the code super easy to follow, thanks! :ok_hand: 

added some comments, mostly on comments and variable names

**Comment by jakobj** [comment]
ah, and also please rebase on the current master, there are some conflicts with the main branch.

**Comment by jakobj** [comment]
ping @HenrikMettler

File: test/test_genome.py (hunk 6 above)
Comment (HenrikMettler) [reply test/test_genome.py:775]: I wasn't fully sure what to do (the logic changed since you wrote that message) - but I have added a few more (complicated) expressions for the output (in `test_set_expression_for_output`)

**Comment by HenrikMettler** [comment]
ping @jakobj

**Review by jakobj — CHANGES_REQUESTED** [review]
almost perfect ;)

i've added a few more comments that should be addressed before merging.

the tests look great, i also can't interpret my old comment any more.

File: cgp/genome.py [hunk 35]
```diff
@@ -270,6 +270,110 @@ def randomize(self, rng: np.random.RandomState) -> None:
...
+        new_dna: List[int]
+            dna segment to be inserted.
+        internal_start_node: int, optional
+            Index of the internal node, where the insert starts.
+            Relative to the first internal node.
+            Defaults to 0.
+
+        Returns
+        ----------
+        address_last_inserted_node: int
```
[the remark this answers is not in the document]
Comment (HenrikMettler) [reply cgp/genome.py:289]: I guess you meant `last_inserted_node: int` ?

**Review by jakobj — APPROVED** [review]
awesome! :rocket: :+1: and merging.

**Comment by jakobj** [comment]
ah, no merging yet: could you please squash the commits? :)

**Comment by jakobj** [comment]
awesome!
